clang/lib/Sema/SemaStmtAttr.cpp

```diff
@@ -322,6 +322,81 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
   return ::new (S.Context) UnlikelyAttr(S.Context, A);
 }
 
+CodeAlignAttr *Sema::BuildCodeAlignAttr(const AttributeCommonInfo &CI,
+                                        Expr *E) {
+  if (!E->isValueDependent()) {
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return nullptr;
+    E = Res.get();
+
+    // This attribute requires an integer argument which is a constant power of
+    // two between 1 and 4096 inclusive.
+    if (ArgVal < CodeAlignAttr::MinimumAlignment ||
+        (ArgVal < CodeAlignAttr::MaximumAlignment && !ArgVal.isPowerOf2())) {
+      Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
+          << CI << CodeAlignAttr::MinimumAlignment
+          << CodeAlignAttr::MaximumAlignment << ArgVal.getSExtValue();
+      return nullptr;
+    }
+
+    if (ArgVal > CodeAlignAttr::MaximumAlignment) {
+      if (ArgVal > std::numeric_limits<int32_t>::max() &&
```

Comment:
This isn't a great solution either.  This results in us lying to users.  Additionally, we have no problem with less than 64 bits.

Also note, you have the same problem with something sufficiently negative as well.  Part of the problem is we only have diagnostic overloads for `<<` up to 'long long'.  I might suggest printing the expression instead if `tryGetSExtValue` fails.

Reply:
>>This isn't a great solution either. This results in us lying to users. 

@erichkeane, do you have any suggestion about how to handle this in a bettier way?

>>Additionally, we have no problem with less than 64 bits.
Correct

>>I might suggest printing the expression instead if tryGetSExtValue fails.

What is `` tryGetSExtValue``?  Could you please elaborate? Thank you

Reply:
`https://llvm.org/doxygen/classllvm_1_1APInt.html#ae30d62aebec681aceb655de2489f12ba`

Reply:
@erichkeane , Thanks for the pointer. I have updated patch with with trySExtValue() and added test case for big negative alignment value: https://github.com/llvm/llvm-project/pull/70762/commits/5ebb42f3105186ffde314496bd9203bca5f18c45